rules/directory/filter.bzl

```diff
@@ -0,0 +1,65 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Rules to filter files from a directory."""
+
+load(":providers.bzl", "DirectoryInfo")
+load(":utils.bzl", "get_relative")
+
+def _directory_filter_impl(ctx):
+    directory = ctx.attr.directory[DirectoryInfo]
+    srcs = depset([
```

Comment:
After my refactor away from depsets, the implementation of this is basically just:
`srcs = ["foo/bar"]` -> `directory.entries["foo"].entries["bar"]`

IMO, this is pretty reasonable, since this allows us to provide nice errors such as distinguishing between "bar doesn't exist within foo" and "the directory foo doesn't exist".